x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/BinaryDateTimeFunction.java

```diff
@@ -0,0 +1,131 @@
+import static org.elasticsearch.xpack.sql.expression.function.scalar.datetime.BinaryDateTimeProcessor.BinaryDateOperation;
+import static org.elasticsearch.xpack.sql.expression.function.scalar.datetime.BinaryDateTimeProcessor.BinaryDateOperation.TRUNC;
+import static org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder.paramsBuilder;
+
+public abstract class BinaryDateTimeFunction extends BinaryScalarFunction {
+
+    private final ZoneId zoneId;
+    private final BinaryDateOperation operation;
+
+    public BinaryDateTimeFunction(Source source, Expression truncateTo, Expression timestamp, ZoneId zoneId,
+                                  BinaryDateOperation operation) {
+        super(source, truncateTo, timestamp);
+        this.zoneId = zoneId;
+        this.operation = operation;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        TypeResolution resolution = isString(left(), sourceText(), Expressions.ParamOrdinal.FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        if (left().foldable()) {
+            String truncateToValue = (String) left().fold();
+            if (truncateToValue != null && resolveDateTimeField(truncateToValue) == false) {
+                List<String> similar = findSimilarDateTimeFields(truncateToValue);
+                if (similar.isEmpty()) {
+                    return new TypeResolution(format(null, "first argument of [{}] must be one of {} or their aliases, found value [{}]",
+                        sourceText(),
+                        validDateTimeFieldValues(),
+                        Expressions.name(left())));
+                } else {
+                    return new TypeResolution(format(null, "Unknown value [{}] for first argument of [{}]; did you mean {}?",
+                        Expressions.name(left()),
+                        sourceText(),
+                        similar));
+                }
+            }
+        }
+        resolution = isDate(right(), sourceText(), Expressions.ParamOrdinal.SECOND);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    public ZoneId zoneId() {
+        return zoneId;
+    }
+
+    protected abstract boolean resolveDateTimeField(String dateTimeField);
+
+    protected abstract List<String> findSimilarDateTimeFields(String dateTimeField);
+
+    protected abstract List<String> validDateTimeFieldValues();
+
+    @Override
+    protected Pipe makePipe() {
+        return new BinaryDateTimePipe(source(), this, Expressions.pipe(left()), Expressions.pipe(right()), zoneId, operation);
+    }
+
+    @Override
+    public Nullability nullable() {
+        return Nullability.TRUE;
+    }
+
+    @Override
+    public Object fold() {
+        if (operation == TRUNC) {
+            return DateTruncProcessor.process(left().fold(), right().fold(), zoneId);
+        } else {
```

Comment:
I would add a comment here or for the entire class pointing out that, for the moment, this class has only two subclasses - datepart and datetrunc - thus, the `else` branch points directly to `DatePartProcessor`.